spinalcordtoolbox/labels.py

```diff
@@ -341,15 +341,12 @@ def remove_missing_labels(img: Image, ref: Image):
     :param ref: reference image
     :returns: image with labels missing from reference removed
     """
-    out = img.copy()
-
-    input_coords = img.getNonZeroCoordinates(coordValue=True)
-    ref_coords = ref.getNonZeroCoordinates(coordValue=True)
-
-    for c in input_coords:
-        if c not in ref_coords:
-            out.data[int(c.x), int(c.y), int(c.z)] = 0
+    out = zeros_like(img)
 
+    for src_coord in img.getNonZeroCoordinates(coordValue=True):
+        for ref_coord in ref.getNonZeroCoordinates(coordValue=True):
+            if src_coord == ref_coord:
+                out.data[src_coord.x, src_coord.y, src_coord.z] = src_coord.value
```

Comment:
If I understand correctly, this is basically "[check for common elements between 2 lists](https://stackoverflow.com/questions/2864842/common-elements-comparison-between-2-lists), then copy them over to an output image," right?

I'm not sure if this even needs optimization, considering it's unlikely there are even _that_ many labels, but, using sets would remove one of the nested loops:

```python
    input_coords = img.getNonZeroCoordinates(coordValue=True)
    ref_coords = ref.getNonZeroCoordinates(coordValue=True)
    common_coords = list(set(input_coords).intersection(ref_coords))

    out = zeros_like(img)
    for coord in common_coords;
        out.data[coord.x, coord.y, coord.z] = coord.value 
```

Reply:
> This PR is fixing 2 issues, right?
> 
>     1. Broken test (because of wrong input data)
> 
>     2. Slow, inefficient test
> 
> 
> If so, it might be good to clarify the issue description, as #3153 only reports 1).

Initial diagnosis was a slow test, but while working on it I realized that it was actually broken. It was also using a wrong input, causing unnecessary slowness.

Reply:
Thanks for the in-depth explanation. Your changes make much more sense now.

I'm wondering if this is more a flaw with with `Coordinate.__eq__()`, then. I'm happy to poke around a little more, but I'll try to approve this soon (by today) as you've already got a working solution and I don't want to hold this up for very long.

Reply:
I dug a little deeper into the `test_sct_register_to_template` failure:

With my comparison, this is the result:

![Screenshot from 2021-02-08 13-00-26](https://user-images.githubusercontent.com/16181459/107262174-35d26780-6a0e-11eb-97a0-beaa3517f958.png)

With your comparison, this is the result:

![Screenshot from 2021-02-08 13-04-18](https://user-images.githubusercontent.com/16181459/107262130-2c48ff80-6a0e-11eb-9901-d933afd8f0ea.png)

So, the coordinates _values_ were being compared correctly (i.e. `__eq__()` worked OK), but the coordinates were taken from the wrong image (wrong x/y/z). The fix for my suggestion is pretty simple:

```diff
-    common_coords = list(set(input_coords).intersection(ref_coords))
+    common_coords = list(set(ref_coords).intersection(input_coords))
```

Still, now that I see how this went wrong, using `set` isn't as clear as I thought it was. So, I think we should just stick with your solution.

Reply:
Thanks for digging! 

It may be worth revisiting if performance needs to be optimized. Your solution has a faster average time complexity than mine.